### DISCUSSION TITLE
Reposerver HA with PVC

### DISCUSSION THREAD

**Discussion by fernandezcuesta** [argoproj/argo-cd#17115] [Q&A]
As mentioned in the [High Availability guide](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-repo-server), repo-server can be mounted a PVC into `/tmp`. By having a look at the [Helm chart](https://github.com/argoproj/argo-helm/blob/argo-cd-5.55.0/charts/argo-cd/templates/argocd-repo-server/deployment.yaml#L419) this is controlled by `.Values.repoServer.existingVolumes.tmp` which is a reference to a single PVC.
However, in HA repo-server has either >1 replicas or HPA, which prevents a `ReadWriteOnce` PV from being mounted by multiple pods.

Is repo-server capable of working with a `ReadWriteMany` persistent volume, especially when it comes to multiple pods writing to the same persistent storage simultaneously?

**Comment by jgwest** [comment]
The `/tmp` folder should definitely not be shared between multiple concurrent processes, as would be the case with a repo server /tmp mounted as a ReadWriteMany volume. The reposerver uses an in-process locking system to ensure that `/tmp` can be safely shared within a single process, but this would break if shared between multiple processes.

**Comment by davidfalconego** [comment]
any idea, im still facing an issue with argocd HA and a PVC pointed to thoses argocd-repo PODS to consume a different NFS disk

**Comment by moore-nathan** [comment]
So far I have seen that when repo-server has a `ReadWriteMany` PVC and multiple replicas it errors out with 
```
Unable to load data: Failed to fetch default: `git fetch origin --tags --force --prune` failed exit status 128: time="2024-05-07T17:49:53Z" level=fatal msg="dial unix /tmp/reposerver-ask-pass.sock: connect: connection refused" error: unable to read askpass response from 'argocd' fatal: could not read Username for 'https://github.com': terminal prompts disabled
```
There is a .sock file that gets created for connections in the `/tmp` file and when sharing among pods this causes problems.
I'm not an expert on this, but this is the behavior I've observed. Would really like to find a workaround for this, and if its not possible this should definitely be mentioned in the docs as a known limitation.

**Comment by vitas** [comment]
there is a pull request to make statefulset, unfortunately PR is stack since 2022

**Comment by ChristianCiach** [comment]
The *correct* solution here is to use "ephemeral volumes", as shown here: https://github.com/argoproj/argo-helm/issues/438#issuecomment-1137635193

Ephemeral volumes allow you to combine the automatic creation of PVs (similar to StatefulSet volumeClaimTemplates) with the stateless nature of `Deployment` resources. Modeling the repo-server as a StatefulSet would make its volumes *persistent*, even though the repo-server's data should only be treated as a cache. So epeheral volumes are the obvious way to go here.

**Reply by vitas** [reply]
Yes, this a good way, unfortunately not all CSI drivers support that, for example trident CSI we use does not

**Comment by inderpartap** [comment]
There is a `TMPDIR` envvar which you could use to change the `/tmp` directory to which repo-server caches, and then mount the volume on that new path. The subPathExpr allows us to have a unique path for every pod. We are connecting a single EFS to all the repo-server pods in the following way - 

```yaml
repoServer
  volumes:
    - name: repo-server-efs-volume
      persistentVolumeClaim:
        claimName: repo-server-efs-pvc
  volumeMounts:
    - name: repo-server-efs-volume
      mountPath: /tmp-efs  # should match what you specify in TMPDIR
      subPathExpr: $(POD_NAME)
```

Let me know if this works for you

**Reply by vitas** [reply]
@inderpartap thank you for the snippet,  how the unused space will be freed, i mean after old pod is destroyed ?
